### PR TITLE
Refactor imports

### PR DESCRIPTION
**Pull request by dnephin**
I believe in order to implement #103 properly 2 changes must be made:
* `types.TypeString` needs to be called with `types.RelativeTo` to remove the package selector when the package matches the target package
* the template needs to use `types.TypeString` for the line with `var _ {{.TargetAlias}}.{{.TargetName}} = `

This refactor should make it easier to implement both of these changes. It reduces the code required to handle imports, and I believe makes it easier to follow by doing de-duplication immediately, instead of a post-processing step, which required changing aliases and logic related to imports to 
Index imports so that any duplicate aliase can be detected immediately before they are added.

Instead of hiding the types.TypeString call, implement a `types.Qualifier` which can be used to return the package alias. In a follow up PR this can be updated to make the call to `types.RelativeTo`.

### CODE REVIEW

File: generator/import.go [hunk 1]
```diff
@@ -1,49 +1,54 @@
...
-		}
-		return f.Imports[i].Path < f.Imports[j].Path
-	})
+	result := Import{Alias: alias, PkgPath: path}
+	i.ByPkgPath[path] = result
+	i.ByAlias[alias] = result
+	return result
 }
 
+// TODO: what version of go required this?
```
Comment (joefitzgerald) [inline generator/import.go]: This is required in go 1.11 and go 1.10 when using `go/packages` to load packages. The files that are loaded from vendor folders will include the vendor path in the package path, making it invalid to use as an import path unless we strip out the vendor prefix.

This code is functionally identical to code used in golang.org/x/tools packages.

File: generator/import.go [hunk 2]
```diff
@@ -1,49 +1,54 @@
...
-	Path  string
+	Alias   string
+	PkgPath string
 }
 
 // AddImport creates an import with the given alias and path, and adds it to
 // Fake.Imports.
-func (f *Fake) AddImport(alias string, path string) Import {
+func (i *Imports) Add(alias string, path string) Import {
+	// TODO: why is there extra whitespace on these args?
```
Comment (joefitzgerald) [inline generator/import.go]: Not sure I understand; which white space / args are you referring to?

Comment (dnephin) [reply generator/import.go]: `path` and `alias` are both passed to `strings.TrimSpace` so I assume there is either leading or trailing whitespace on those strings, but it is not clear to me where that would come from. It seems like this is called with values from inspecting `go/types`, which I would assume do not contain extra whitespace.

Comment (joefitzgerald) [reply generator/import.go]: I guess I am just always defensive with function arguments (particularly when exported).

**Pull request merged**